Calculator [STEP2] Mason, Blue

qwerty3345 opened the pull request:
안녕하세요 라자냐! @wonhee009
팀 🔮MABLE🔮 의 Mason, Blue 입니다 😊
아직 아쉬운 점이 많지만, STEP 2에서 제시한 요구사항을 최대한 준수하고,
더 앞서나가지 않는 선에서 프로젝트를 진행하였습니다.
설 연휴동안 맛있는 음식 많이 드시고 새해 복 많이 받으세요! 🙇
리뷰 잘 부탁드립니다! 감사합니다 💙

# ✅ 요구사항 & 체크리스트

- 제시된 UML대로 연산과 관련된 타입 구현
    - [x]  enum Operator
    - [x]  extension String
    - [x]  extension Double
    - [x]  enum ExpressionParser
    - [x]  struct Formula
- [x]  고차함수 활용
    - compactMap, flatMap, reduce
- 예외처리
    - [x]  0으로 나누기를 시도하는 경우 적절한 오류처리
    - 각종 예외사항과 오류사항에 대해 대처할 수 있으면 좋음
        - [x]  피연산자와 연산자의 개수가 잘못된 경우
        - [x]  연산할 요소가 큐에 존재하지 않는 경우

# 💬 UML

- ⭐️ 해당 UML은 프로젝트 기능 명세로, 수정이 불가한 요구사항이었습니다.


![](https://user-images.githubusercontent.com/70311145/188132535-8bc81913-7863-4877-88f9-0f8e811a52bd.jpeg)



<h1 style="caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0); font-style: normal; font-variant-caps: normal; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-size-adjust: auto; -webkit-text-stroke-width: 0px; text-decoration: none;">📝 객체별 역할과 책임</h1>

Name | Type | 역할과 책임
-- | -- | --
Operator | enum | 각 연산자 (+, -, ÷, ×) 케이스, 케이스별 연산 로직
CalculateItem | protocol | 연산 가능 요소를 제한하기 위한 protocol
ExpressionParser | enum | 연산정보인 input 문자열을 받아 파싱 해 Formula 로 변환함
Formula | enum | 연산자, 피연산자 Queue들을 갖고, 해당 요소들의 연산을 수행할 수 있는 객체
CalculatorItemQueue | struct | 연산자, 피연산자를 저장하는 Queue 자료구조 (링크드 리스트 구조로 구현)
LinkedList | struct | 노드간의 연결로 구현한 링크드리스트 자료구조
Node | class | 링크드리스트 내부의 노드
CalculatorError | enum | 계산기에서 발생할 수 있는 오류 정의 및 Description 메세지 정의

# ✏️ 학습 키워드

- 고차함수
- 제네릭, Protocol
- UML
- Unit Test

# **⁉️ 고민과 해결**

## 에러 케이스의 네이밍

0으로 나누는 경우의 오류 처리를 위해 Error 타입을 정의한 후 필요하다고 생각되는 에러 케이스들을 추가하였습니다.
에러케이스 중 피 연산자와 연산자의 개수가 일치하지 않는 경우 오류를 반환하는 케이스를 추가하려고 네이밍을 하는 과정에 많은 시간을 할애하게 되었습니다.

```swift
//CalculatorError.swift
enum CalculatorError: LocalizedError {
    case divideByZero
    case notMatchingCountOfOperatorsAndOperands
    case invalidDequeue

    var errorDescription: String? {
        switch self {
        case .divideByZero:
            return "NaN"
        case .notMatchingCountOfOperatorsAndOperands:
            return "피연산자와 연산자의 개수가 잘못되었습니다."
        case .invalidDequeue:
            return "연산할 요소가 존재하지 않습니다."
        }
    }
}
```

- 다른 에러 케이스처럼 짧고 명확하게 적고 싶었으나 적절한 네이밍이 떠오르지 않아, 명확성을 지키는 방법을 선택하였습니다.
- `notMatchingCountOfOperatorsAndOperands` 라는 네이밍이 마음에 들진 않았지만, 다른 네이밍이 떠오르지 않았습니다.

> **Formula.swift**의 `result()` 메서드 수행 로직
> 
> 1. result() 함수 실행
> 2. 피연산자에 요소가
>     - 있는 경우: 첫 피연산자를 dequeue
>     - 없는 경우: `invalidDequeue` error return
> 3. 피연산자 큐의 요소 개수와 연산자 큐의 요소 개수가
>     - 같으면: 다음 단계 진행
>     - ⭐ 다르면: `notMatchingCountOfOperatorsAndOperands` error return
> 4. 연산자와 피연산자를 dequeue한 값이
>     - `nil`이라면: invalidDequeue error return
>     - `nil`이 아니라면: 다음 단계 진행
> 5. 0으로 나누는 연산이
>     - 있는 경우: `divideByZero` error return
>     - 없는 경우: 연산 수행

🤔❓ 저희의 에러 케이스 네이밍에 대한 라자냐의 의견은 어떠신가요? 다른 테스트 케이스와 동일하게 간단하게 작성하는 것이 좋을까요?

## 파싱 로직 내부의 split 함수 활용

저희가 UML 을 분석 한 결과는 아래와 같습니다.

1. input이 하나의 String 값으로 들어온다. (”1+2÷9”…)
2. 해당 연산 String 을 연산자, 피연산자(숫자)들로 파싱하고 각각 큐에 담아 Formula 를 반환한다.
1. 해당 Formula 의 result 를 호출하여 각각의 큐에 담긴 연산자, 피연산자를 dequeue 하여 계산하고 결과값을 반환함.

이 때 2번의 파싱 로직에 있어서, 저희가 생각한 로직은 

- 연산자는 enum 의 rawValue 로 생성하여 compactMap 으로 배열화 해주고,
- 피연산자(숫자)는 각각의 연산자를 기준으로 잘라서 배열화 한다.

이 때 처음 구상한 피연산자(숫자) 파싱 로직은 각각의 연산자 Character 를 Set으로 담고, 문자열을 한 번에 해당 연산자들을 기준으로 components 로 잘라주는 로직이었습니다.

> 처음에 구상한 숫자 파싱 로직
> 

```swift
private static func componentsByOperaters(from input: String) -> [String] {
    let operatorCharacters = Operator.allCases.map { $0.rawValue }
    let operatorCharacterSet = CharacterSet()
    return input.components(separatedBy: operatorCharacterSet)
}
```

그런데, UML 기능명세서의 String extension 의 split 메서드를 사용하여 하려니, 
아래처럼 각각의 요소마다 split을 하고 flatMap을 해주는 방식으로 구현해야 했습니다.

처음 구상한 로직보다 비효율적인 로직처럼 보였지만… 명세서 UML의 구조를 최대한 지키려 이렇게 구현하였습니다. 

> 최종 구현한 숫자 파싱 로직
> 

```swift
extension String {
		// UML 기능 명세에 있는 함수
    func split(with target: Character) -> [String] {
        return components(separatedBy: String(target))
    }
}

// enum ExpressionParser 로직
private static func componentsByOperaters(from input: String) -> [String] {
    return Operator.allCases.reduce([input]) { partialResult, calculatorOperator in
        partialResult.flatMap { $0.split(with: calculatorOperator.rawValue) }
    }
}
```

❓🤔 이런 문제에 대한 라자냐의 의견은 어떠신지 궁금합니다.

qwerty3345 committed:
feat: Formula, ExpressionParser dummy Type 생성

calledBlu committed:
feat: 연산자 및 연산 수행 메서드를 포함하는 열거형 Operator 구현

qwerty3345 committed:
refactor: step1-fix 에서 Queue 의 속성 이름 front, rear 로 수정한 내용 merge

qwerty3345 committed:
Merge branch 'step2' of github.com:qwerty3345/ios-calculator into step2

calledBlu committed:
feat: String extension을 통해 split(with target:) 메서드 확장

qwerty3345 committed:
chore: .gitignore에 .DS_Store 추가

calledBlu committed:
chore: 불필요한 indent 삭제

qwerty3345 committed:
feat: LinkedList, Queue 갯수를 반환하는 count 속성 추가

qwerty3345 committed:
feat: Formula result 계산 로직 구현 (테스트 전)

calledBlu committed:
test: Formula의 result() 메서드 테스트 작성 및 확인

qwerty3345 committed:
refactor: Formula 내부의 queue 프로퍼티 은닉화

calledBlu committed:
feat: Formula를 반환하는 parse(from input: ) 메서드 생성

qwerty3345 committed:
feat: componentsByOperators 메서드 구현 및 pase 메서드 리팩터링

- parse 메서드에서 Queue 생성 시 이니셜라이저로 요소들을 받아서 생성하도록 로직 변경

calledBlu committed:
test: ExpressionParser 내의 parse 테스트 실행

qwerty3345 committed:
refactor: for문 로직 reduce 고차함수 로직으로 리팩터링

calledBlu committed:
refactor: Operator 내의 rawValue 연산자를 UILabel과 동일하게 수정

qwerty3345 committed:
test: ExpressionParser의 parse메서드가 산출한 Fomula의 result가 일치하는 값을 계산하는지 테스트

calledBlu committed:
feat: CalculatorError 에러 타입 정의

- divideByZero: 0으로 나누는 경우
- notMatchingCountOfOperatorsAndOperands: 연산자와 피연산자의 개수가 일치하지 않는 경우

qwerty3345 committed:
refactor: 정의한 에러 타입에 대한 예외 상황 처리

calledBlu committed:
refactor: 새로운 에러 케이스 추가 및 리팩토링

- invalidDequeue: 연산할 요소가 없어서 dequeue를 할 수 없는 경우

qwerty3345 committed:
test: ExpressionParser 테스트 케이스 추가

- test_연산자와_피연산자_갯수가_적절하지_않은경우_에러를_반환한다
- test_0으로_나눌시_에러를_반환한다

qwerty3345 committed:
chore: CalculateItem 프로토콜 위치 변경 및 파일 삭제

calledBlu committed:
refactor: final, private 키워드 추가

qwerty3345 committed:
chore: 고차함수 내부 줄 바꿈 처리

qwerty3345 committed:
docs: STEP2까지의 리드미 작성

wonhee009 commented on Calculator/CalculateItemQueueTest/CalculatorItemQueueTest.swift:
```diff
@@ -21,48 +21,85 @@ final class CalculatorItemQueueTest: XCTestCase {
+        }
+
+        // when
+        sut.dequeue()
+        sut.dequeue()
```
`inputs`의 count만큼 `dequeue()`하도록 하면 inputs의 갯수가 변하더라도 해당 테스트에 대한 검증은 보장될거 같습니다.

qwerty3345 replied:
수정 반영 하였습니다!

[be94391](https://github.com/tasty-code/ios-calculator/pull/22/commits/be9439159abb0cda60a6338ce1b4a45aa2f2d5a3)

calledBlu replied:
inputs의 count만큼 `dequeue()`하도록 수정하였습니다!

> 해당 커밋: ([be94391](https://github.com/tasty-code/ios-calculator/pull/22/commits/be9439159abb0cda60a6338ce1b4a45aa2f2d5a3))

wonhee009 commented on Calculator/Calculator/Model/CalculateItem.swift:
```diff
@@ -6,7 +6,3 @@
 //
 
 import Foundation
-
-protocol CalculateItem {
-    
-}
```
해당 파일은 삭제 안해도 되는건가요?.?

qwerty3345 replied:
Remove Trash 가 아닌, Remove Reference 를 해서 생긴 문제였던 것 같습니다!
두 삭제의 차이에 대해 학습했고, 삭제 해 주었습니다 :)
> 해당 커밋: [349b307](https://github.com/tasty-code/ios-calculator/pull/22/commits/349b307010ff2e07c87724ca8d8ed858d6640bdf)

wonhee009 commented on Calculator/Calculator/Model/ExpressionParser.swift:
```diff
@@ -0,0 +1,28 @@
+        return Formula(operands: operands, operators: operators)
+    }
+
+    private static func componentsByOperaters(from input: String) -> [String] {
```
```suggestion
    private static func componentsByOperand(from input: String) -> [String] {
```
아닌가요?.?

calledBlu replied:
해당 프로젝트에서 제시된 UML에 따라 함수명을 작성한 것입니다!
PR 메세지에 첨부드린 UML을 보시면 componentsByOperaters(from input:) -> [String]으로 되어있습니다! 😄

wonhee009 commented on Calculator/ExpressionParserTests/ExpressionParserTests.swift:
```diff
@@ -0,0 +1,99 @@
+        XCTAssertTrue(expectationOperands.allSatisfy { $0 == formula.operands.dequeue() })
+    }
+
+    func test_parse메서드로_산출한_Fomula의_result가_일치하는_값이다() {
```
```suggestion
    func test_parse메서드로_산출한_Formula의_result가_일치하는_값이다() {
```

qwerty3345 replied:
오타 지적 감사합니다! 
> 해당 커밋: [368309b](https://github.com/tasty-code/ios-calculator/pull/22/commits/368309bfbe24bd1b0d318c677e01e8f11523195c)

wonhee009 commented on Calculator/ExpressionParserTests/ExpressionParserTests.swift:
```diff
@@ -0,0 +1,99 @@
+
+    func test_parse메서드로_산출한_Fomula의_result가_일치하는_값이다() {
+        // given
+        let input = "1\(addSign)0\(subtractSign)10\(multiplySign)5.5\(divideSign)5.5\(addSign)21\(subtractSign)1.1"
+        let expectation = 10.9
```
```swift
let inputNubmer = [1, 0, 10, 5.5, 5.5, 21, 1.1]
let inputSign = [+, -, ...]
```
이런식으로 프로퍼티를 정의하고 for문을 통해 expectation을 구하는건 어떤가요?
지금 코드라면 input의 sign이나 숫자가 수정될 경우 expectation도 같이 갱신해줘야할거 같은데 그 과정에서 휴먼에러가 있을 수 있을거 같습니다.

calledBlu replied:
라자냐의 추천대로 휴먼에러 발생 가능성을 낮추기 위해 테스트 케이스를 수정해 보았습니다.

`ExpressionParser`가 계산을 하는 역할이 아닌 연산자를 기준으로 input 문자열을 파싱하고, `Formula`를 생성하는 역할을 수행합니다.

이를 테스트하기 위해, Formula가 제대로 생성되었는지를 판단할 수 있도록 연산 결과인 expectation을 도출하기 위한 함수 `generateInput()`와 파싱 대상이 될 문자열을 생성하는 함수`calculate()`를 생성하여 진행을 해보려고 했습니다.
> ExpressionParser 테스트 수정 커밋: ([5bbc3bd](https://github.com/tasty-code/ios-calculator/pull/22/commits/5bbc3bdef735add67ce8ce5167261ac029c71978))


하지만, 수정 과정에서 해당 테스트 케이스의 역할이 `ExpressionParser`를 테스트하는 데 부적합하다고 판단하여 **삭제**하였고, 라자냐가 추천해주신 방식을 `Formula`를 테스트할 때 적용해 보았습니다.
> Formula 테스트 수정 커밋: ([3f220d7](https://github.com/tasty-code/ios-calculator/pull/22/commits/3f220d75203eed954a72e66b1c27fd536b2e34ed))

wonhee009 reviewed (changes requested):
수고하셨습니다 :)
코멘트 남겨두었으니 확인 부탁드려요.

- 에러 네이밍이 명확한거 좋습니다~
- 문제라는게 어떤걸 말하는걸까요?.?

qwerty3345 committed:
chore: CalculateItem 파일 삭제

Xcode 에서 Remove Reference 로 삭제하여 로컬에 반영되지 않았었음
- Move To Trash 로 삭제 진행

calledBlu committed:
refactor: input의 count만큼 dequeue() 하도록 테스트 케이스 수정

qwerty3345 committed:
chore: Fomula -> Formula 오타 수정

qwerty3345 committed:
refactor: ExpressionParserTest 휴먼에러가 발생할 수 있는 테스트 케이스 수정

- test_parse메서드로_산출한_Formula의_result가_일치하는_값이다 내부의 expectation 을 산출하는 방식을 직접 계산에서 함수가 계산하게 변경함

qwerty3345 committed:
refactor: 테스트 로직 변경

- ExpressionParserTest 에 존재하던 계산 테스트를 삭제 (Formula 에서 수행)
- FormulaTest 의 계산 테스트의 휴먼에러를 방지하기 위한, 계산 메서드 추가

qwerty3345 commented:
> 문제라는게 어떤걸 말하는걸까요?.?

질문드린 것처럼 파싱을 수행하는 로직을 작성할 때, 
**"기능 명세서의 메서드를 정확히 활용하기"** vs **"더 효율적이어 보이는 로직을 활용하기"**
같은 문제에서 라자냐라면 어떤 것을 선택하실 것 같은지... 대해 질문드리고 싶었습니다.

wonhee009 reviewed (approved):
수고하셨습니다~

저라면 명세서를 따랐을거 같습니다. 실제로 프로젝트를 진행하면서 명세서를 작성할때에는 다른 개발자들과 합의 하에 작성 혹은 유지되기 때문에 다른 방식으로 구현하고 싶다면 사전에 논의를 거쳐야할거 같습니다.
